### DISCUSSION TITLE
Expose Row Index in PixelRowOperation delegate

### DISCUSSION THREAD

**Discussion by PeteMichaud** [SixLabors/ImageSharp#2055] [Ideas]
I've been working through Joe's answer to this question: https://stackoverflow.com/questions/71388492/right-way-to-parallelize-pixel-access-across-multiple-images-using-imagesharp/71388733#71388733

I am trying to take his recommendation seriously not to used the `Advanced` stuff if possible, but I really do need parallel processing as the images I'm processing regularly push dimensions in the range of 25-50k width*height, so I've been looking into the Vector4 api he mentioned, which is given [here](https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html) as an example: 

```
// from https://docs.sixlabors.com/articles/imagesharp/pixelbuffers.html
image.Mutate(c => c.ProcessPixelRowsAsVector4(row =>
{
    for (int x = 0; x < row.Length; x++)
    {
        // We can apply any custom processing logic here
        row[x] = Vector4.SquareRoot(row[x]);
    }
}));
```

I can't use this because my calculations absolutely require the current y/rowIndex coordinate. I looked [here](https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Processing/PixelRowOperation.cs) and found that one version of the delegate takes a `T value` arg, but as far as I can tell, that can't get me the index (maybe I'm wrong here?)

My suggestion is simply to provide an overload that has an index like so: 

```
image.Mutate(c => c.ProcessPixelRowsAsVector4((rowIndex, row) =>
{
    for (int x = 0; x < row.Length; x++)
    {
        // y = rowIndex
        // x = x
    }
}));
```
I assume it doesn't make sense to attach an `index` type property to the row `Span` object itself, but that's a different possibility if the Span is less generic than I'm assuming it is.

**Comment by antonfirsov** [comment]
Sounds reasonable to me, are you interested creating a PR?

**Reply by PeteMichaud** [reply]
The spirit is willing, but I have very little confidence that I can navigate this codebase within a reasonable timeframe in a way that ends up with an acceptable-to-the-team PR.   

One thing I'll note is that I investigated the code around this idea, and found a bunch of overloads for `ProcessPixelRowsAsVector4` [here](https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Processing/Extensions/Effects/PixelRowDelegateExtensions.cs)

And one in particular seemed promising: https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Processing/Extensions/Effects/PixelRowDelegateExtensions.cs#L100

Because it creates something called a `PositionAwarePixelRowDelegateProcessor` which sounds like the sort of thing we're after. 

[That class](https://github.com/SixLabors/ImageSharp/blob/e54219acec10d2d2dc247ed5fd4230204a89c07b/src/ImageSharp/Processing/Processors/Effects/PositionAwarePixelRowDelegateProcessor.cs#L14) does provide an Invoke signature that includes a `Point offset` which is maybe equivalent to the functionality we're talking about here, and might be the right place to start to provide the more clear `(index, row)` delegate signature we're talking about.

**Comment by JimBobSquarePants** [comment]
As you've discovered there's an overload that passes the current position as a  `Point` of which `Y` is the current row so you have access to that functionality without any code changes. 

We won't be adding additional overloads as to do so actually involves a lot of work with no added value.
